apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfSubscriber.cs

```diff
@@ -0,0 +1,723 @@
+                GaxPreconditions.CheckState(!_globalSoftStopCts.IsCancellationRequested, "Can only stop an instance once.");
+                _globalSoftStopCts.Cancel();
+            }
+            var registration = hardStopToken.Register(() => _globalHardStopCts.Cancel());
+            // Do not register this Task, it completes *after* mainTask
+            _taskHelper.Run(async () => await _taskHelper.ConfigureAwaitWithFinally(_mainTcs.Task, () => registration.Dispose()));
+            return _mainTcs.Task;
+        }
+
+        private class Flow
+        {
+            public Flow(long maxByteCount, long maxElementCount, Action<Task> registerTaskFn, TaskHelper taskHelper)
+            {
+                _maxByteCount = maxByteCount;
+                _maxElementCount = maxElementCount;
+                _registerTaskFn = registerTaskFn;
+                _taskHelper = taskHelper;
+            }
+
+            private readonly object _lock = new object();
+            private readonly long _maxByteCount;
+            private readonly long _maxElementCount;
+            private readonly Action<Task> _registerTaskFn;
+            private readonly TaskHelper _taskHelper;
+
+            private long _byteCount;
+            private long _elementCount;
+
+            private TaskCompletionSource<int> _event = new TaskCompletionSource<int>();
+
+            private bool IsFlowOk() => _byteCount < _maxByteCount && _elementCount < _maxElementCount;
+
+            public async Task Process(long byteCount, Func<Task> fn)
+            {
+                Task prevEventTask = null;
+                while (true)
+                {
+                    Task eventTask;
+                    lock (_lock)
+                    {
+                        if (IsFlowOk())
+                        {
+                            _byteCount += byteCount;
+                            _elementCount += 1;
+                            break;
+                        }
+                        eventTask = _event.Task;
+                    }
+                    prevEventTask = eventTask;
+                    await _taskHelper.ConfigureAwait(eventTask);
+                }
+                _registerTaskFn(_taskHelper.Run(async () => await _taskHelper.ConfigureAwaitWithFinally(fn(), () =>
+                {
+                    lock (_lock)
+                    {
+                        bool preIsFlowOk = IsFlowOk();
+                        _byteCount -= byteCount;
+                        _elementCount -= 1;
+                        if (!preIsFlowOk && IsFlowOk())
+                        {
+                            var ev = _event;
+                            _event = new TaskCompletionSource<int>();
+                            ev.SetResult(0);
+                        }
+                    }
+                })));
+            }
+        }
+
+        private class SingleChannel
+        {
+            public SingleChannel(SubscriptionName subscriptionName,
```

Comment:
If this took a HiPerfSubscriberImpl as a parameter, it could get to a lot of these from that (as it's a nested class). Or we could move them into a separate class.